ssl/tls13_enc.c

```diff
@@ -764,12 +772,132 @@ int tls13_change_cipher_state(SSL *s, int which)
+    }
+
+    bio = s->wbio;
+
+    if (!ossl_assert(bio != NULL)) {
+        SSLfatal(s, SSL_AD_INTERNAL_ERROR, SSL_F_TLS13_CHANGE_CIPHER_STATE,
+                 ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+
+    /* All future data will get encrypted by ktls. Flush the BIO or skip ktls */
+    if (BIO_flush(bio) <= 0)
+        goto skip_ktls;
+
+    /* ktls doesn't support renegotiation */
+    if (BIO_get_ktls_send(bio)) {
+        SSLfatal(s, SSL_AD_NO_RENEGOTIATION, SSL_F_TLS13_CHANGE_CIPHER_STATE,
+                 ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    memset(&crypto_info, 0, sizeof(crypto_info));
+    switch (EVP_CIPHER_nid(cipher))
+    {
+#   ifdef OPENSSL_KTLS_AES_GCM_128
+    case NID_aes_128_gcm:
+        crypto_info.gcm128.info.cipher_type = TLS_CIPHER_AES_GCM_128;
+        crypto_info.gcm128.info.version = s->version;
+        crypto_info.tls_crypto_info_len = sizeof(crypto_info.gcm128);
+        memcpy(crypto_info.gcm128.iv, iv + EVP_GCM_TLS_FIXED_IV_LEN,
+            TLS_CIPHER_AES_GCM_128_IV_SIZE);
+        memcpy(crypto_info.gcm128.salt, iv, TLS_CIPHER_AES_GCM_128_SALT_SIZE);
+        memcpy(crypto_info.gcm128.key, key, EVP_CIPHER_key_length(cipher));
+        memcpy(crypto_info.gcm128.rec_seq, &s->rlayer.write_sequence,
+            TLS_CIPHER_AES_GCM_128_REC_SEQ_SIZE);
+        break;
+#   endif
+#   ifdef OPENSSL_KTLS_AES_GCM_256
+    case NID_aes_256_gcm:
+        crypto_info.gcm256.info.cipher_type = TLS_CIPHER_AES_GCM_256;
+        crypto_info.gcm256.info.version = s->version;
+        crypto_info.tls_crypto_info_len = sizeof(crypto_info.gcm256);
+
+        memcpy(crypto_info.gcm256.iv, iv + EVP_GCM_TLS_FIXED_IV_LEN,
+            TLS_CIPHER_AES_GCM_256_IV_SIZE);
+        memcpy(crypto_info.gcm256.salt, iv, TLS_CIPHER_AES_GCM_256_SALT_SIZE);
+        memcpy(crypto_info.gcm256.key, key, EVP_CIPHER_key_length(cipher));
+        memcpy(crypto_info.gcm256.rec_seq, &s->rlayer.write_sequence,
+            TLS_CIPHER_AES_GCM_256_REC_SEQ_SIZE);
+        break;
+#   endif
+#   ifdef OPENSSL_KTLS_AES_CCM_128
+    case NID_aes_128_ccm:
+        crypto_info.ccm128.info.cipher_type = TLS_CIPHER_AES_CCM_128;
+        crypto_info.ccm128.info.version = s->version;
+        crypto_info.tls_crypto_info_len = sizeof(crypto_info.ccm128);
+
+        memcpy(crypto_info.ccm128.iv, iv + EVP_CCM_TLS_FIXED_IV_LEN,
+            TLS_CIPHER_AES_CCM_128_IV_SIZE);
+        memcpy(crypto_info.ccm128.salt, iv, TLS_CIPHER_AES_GCM_128_SALT_SIZE);
+        memcpy(crypto_info.ccm128.key, key, EVP_CIPHER_key_length(cipher));
+        memcpy(crypto_info.ccm128.rec_seq, &s->rlayer.write_sequence,
+            TLS_CIPHER_AES_GCM_128_REC_SEQ_SIZE);
+        break;
+#   endif
+    default:
+        goto skip_ktls;
+    }
+
+    /* ktls works with user provided buffers directly */
+    if (BIO_set_ktls(bio, &crypto_info, which & SSL3_CC_WRITE)) {
+        ssl3_release_write_buffer(s);
+        SSL_set_options(s, SSL_OP_NO_RENEGOTIATION);
```

Comment:
Irrelevant in TLSv1.3

Reply:
Sure, will fix it

Reply:
removed `SSL_set_options(s, SSL_OP_NO_RENEGOTIATION);`